File: code/packages/dart/lz78/lib/src/lz78.dart
import 'dart:typed_data';

/// One LZ78 token represented as `(dictIndex, nextChar)`.
///
/// `dictIndex` points at the longest dictionary prefix that matches the
/// current input. `nextChar` is the following byte, or `0` when a flush token
/// terminates a partial match at end of stream.
class Token {
  /// Creates a token with the given LZ78 fields.
  const Token(this.dictIndex, this.nextChar);

  /// Dictionary entry ID for the longest prefix match, or `0` for a literal.
  final int dictIndex;

  /// Byte following the matched prefix, or the flush sentinel `0`.
  final int nextChar;

  @override
  bool operator ==(Object other) {
    return other is Token &&
        other.dictIndex == dictIndex &&
        other.nextChar == nextChar;
  }

  @override
  int get hashCode => Object.hash(dictIndex, nextChar);

  @override
  String toString() {
    return 'Token(dictIndex: $dictIndex, nextChar: $nextChar)';
  }
}

/// Creates a [Token] without needing to call the constructor directly.
Token token(int dictIndex, int nextChar) => Token(dictIndex, nextChar);

/// Internal trie node used by [TrieCursor].
class _CursorNode {
  _CursorNode(this.dictId);

  final int dictId;
  final Map<int, _CursorNode> children = <int, _CursorNode>{};
}

/// A byte-at-a-time trie cursor for streaming dictionary algorithms.
///
/// The encoder keeps the cursor on the current matched prefix. When [step]
/// fails for a byte, the encoder emits a token, optionally [insert]s the new
/// sequence, and [reset]s to the trie root.
class TrieCursor {
  /// Creates an empty trie cursor positioned at the root.
  TrieCursor()
      : _root = _CursorNode(0),
        _current = _CursorNode(0) {
    _current = _root;
  }

  final _CursorNode _root;
  late _CursorNode _current;

  /// Tries to follow the edge for [byte] from the current position.
  ///
  /// Returns `true` and advances the cursor if the edge exists. Returns
  /// `false` and leaves the cursor unchanged on a miss.
  bool step(int byte) {
    final child = _current.children[byte];
    if (child == null) {
      return false;
    }

    _current = child;
    return true;
  }

  /// Inserts a child edge for [byte] with the given [dictId].
  void insert(int byte, int dictId) {
    _current.children[byte] = _CursorNode(dictId);
  }

  /// Resets the cursor back to the trie root.
  void reset() {
    _current = _root;
  }

  /// Dictionary ID at the current cursor position.
  int get dictId => _current.dictId;

  /// Whether the cursor is currently at the root.
  bool get atRoot => identical(_current, _root);
}

/// Encodes [data] into an LZ78 token stream.
///
/// The encoder walks an explicit trie one byte at a time. On a miss, it emits
/// the current dictionary prefix plus the unmatched byte, records the new
/// sequence in the dictionary if space allows, and resets to the root.
List<Token> encode(Uint8List data, [int maxDictSize = 65536]) {
  final cursor = TrieCursor();
  var nextId = 1;
  final tokens = <Token>[];

  for (final byte in data) {
    if (!cursor.step(byte)) {
      tokens.add(token(cursor.dictId, byte));

      if (nextId < maxDictSize) {
        cursor.insert(byte, nextId);
        nextId += 1;
      }

      cursor.reset();
    }
  }

  if (!cursor.atRoot) {
    tokens.add(token(cursor.dictId, 0));
  }

  return List<Token>.unmodifiable(tokens);
}

/// Decodes an LZ78 token stream back into the original bytes.
///
/// If [originalLength] is provided, the decoded output is truncated to that
/// many bytes so the flush sentinel is not returned to the caller.
Uint8List decode(List<Token> tokens, [int originalLength = -1]) {
  final table = <({int parentId, int byte})>[(parentId: 0, byte: 0)];
  final output = <int>[];

  for (final current in tokens) {
    _validateDecodedToken(current, table.length);

    final sequence = _reconstruct(table, current.dictIndex);
    output.addAll(sequence);
    if (originalLength >= 0 && output.length > originalLength) {
      throw FormatException(
        'Malformed LZ78 token stream: decoded output exceeds declared length $originalLength.',
      );
    }

    if (originalLength < 0 || output.length < originalLength) {
      output.add(current.nextChar);
    } else if (current.nextChar != 0) {
      throw FormatException(
        'Malformed LZ78 token stream: token data exceeds declared length $originalLength.',
      );
    }

    table.add((parentId: current.dictIndex, byte: current.nextChar));
  }

  if (originalLength >= 0 && output.length != originalLength) {
    throw FormatException(
      'Malformed LZ78 token stream: decoded output length ${output.length} does not match declared length $originalLength.',
    );
  }

  return Uint8List.fromList(output);
}

List<int> _reconstruct(List<({int parentId, int byte})> table, int index) {
  if (index == 0) {
    return const <int>[];
  }

  final reversed = <int>[];
  var current = index;
  while (current != 0) {
    final entry = table[current];
    reversed.add(entry.byte);
    current = entry.parentId;
  }

  return reversed.reversed.toList(growable: false);
}

/// Serialises [tokens] to the fixed-width CMP01 wire format.
///
/// Layout:
/// - 4 bytes: original length as big-endian uint32
/// - 4 bytes: token count as big-endian uint32
/// - N x 4 bytes: `(dictIndex: uint16, nextChar: uint8, reserved: uint8)`
Uint8List serialiseTokens(List<Token> tokens, int originalLength) {
  final bytes = ByteData(8 + tokens.length * 4);
  bytes.setUint32(0, originalLength, Endian.big);
  bytes.setUint32(4, tokens.length, Endian.big);

  for (var index = 0; index < tokens.length; index++) {
    final base = 8 + index * 4;
    final current = tokens[index];
    bytes.setUint16(base, current.dictIndex, Endian.big);
    bytes.setUint8(base + 2, current.nextChar);
    bytes.setUint8(base + 3, 0);
  }

  return bytes.buffer.asUint8List();
}

/// Deserialises the CMP01 wire format back into tokens and original length.
({List<Token> tokens, int originalLength}) deserialiseTokens(Uint8List data) {
  if (data.length < 8) {
    throw const FormatException(
      'Malformed LZ78 token stream: header is incomplete.',
    );
  }

  final view = ByteData.view(
    data.buffer,
    data.offsetInBytes,
    data.lengthInBytes,
  );
  final originalLength = view.getUint32(0, Endian.big);
  final tokenCount = view.getUint32(4, Endian.big);
  final expectedLength = 8 + tokenCount * 4;
  if (data.length != expectedLength) {
    throw FormatException(
      'Malformed LZ78 token stream: expected $expectedLength bytes, got ${data.length}.',
    );
  }
  final tokens = <Token>[];

  for (var index = 0; index < tokenCount; index++) {
    final base = 8 + index * 4;
    final current = token(
      view.getUint16(base, Endian.big),
      view.getUint8(base + 2),
    );
    _validateDeserialisedToken(current, originalLength);
    tokens.add(current);
  }

  return (
    tokens: List<Token>.unmodifiable(tokens),
    originalLength: originalLength,
  );
}

/// Compresses [data] using LZ78 and serialises to the CMP01 wire format.
Uint8List compress(Uint8List data, [int maxDictSize = 65536]) {
  return serialiseTokens(encode(data, maxDictSize), data.length);
}

/// Decompresses bytes produced by [compress].
Uint8List decompress(Uint8List data) {
  final decoded = deserialiseTokens(data);
  return decode(decoded.tokens, decoded.originalLength);
}

void _validateDecodedToken(Token current, int dictionaryLength) {
  if (current.dictIndex < 0) {
    throw FormatException(
      'Malformed LZ78 token: dictIndex must be non-negative, got ${current.dictIndex}.',
    );
  }
  if (current.dictIndex >= dictionaryLength) {
    throw FormatException(
      'Malformed LZ78 token: dictIndex ${current.dictIndex} exceeds dictionary size ${dictionaryLength - 1}.',
    );
  }
  if (current.nextChar < 0 || current.nextChar > 255) {
    throw FormatException(
      'Malformed LZ78 token: nextChar must be in 0..255, got ${current.nextChar}.',
    );
  }
}

void _validateDeserialisedToken(Token current, int originalLength) {
  if (originalLength == 0 && current.dictIndex != 0) {
    throw const FormatException(
      'Malformed LZ78 token stream: non-empty dictionary reference for empty output.',
    );
  }
}

File: code/packages/haskell/lz78/src/Lz78.hs
-- | LZ78 explicit-dictionary compression from CMP01.
--
-- The encoder and decoder independently build the same dictionary. The wire
-- representation stores the original length, a token count, and fixed-width
-- four-byte token records.
module Lz78
    ( Token (..)
    , Lz78Error (..)
    , TrieCursor
    , defaultMaxDictionarySize
    , emptyCursor
    , stepCursor
    , insertCursor
    , resetCursor
    , cursorDictId
    , cursorAtRoot
    , encode
    , decode
    , serialiseTokens
    , deserialiseTokens
    , compress
    , compressDefault
    , decompress
    ) where

import Data.Bits ((.|.), shiftL, shiftR)
import qualified Data.ByteString as BS
import Data.ByteString (ByteString)
import qualified Data.IntMap.Strict as IntMap
import Data.IntMap.Strict (IntMap)
import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map)
import Data.Word (Word16, Word32, Word8)

-- | One LZ78 token. Index zero denotes the empty dictionary prefix.
data Token = Token
    { dictIndex :: Word16
    , nextChar :: Word8
    }
    deriving (Eq, Show)

-- | Structural failures detected while reading or decoding a token stream.
data Lz78Error
    = HeaderTooShort Int
    | WireLengthMismatch Int Int
    | NonZeroReservedByte Int Word8
    | InvalidDictionaryIndex Word16 Int
    | InvalidOriginalLength Int
    | DecodedLengthMismatch Int Int
    deriving (Eq, Show)

data TrieNode = TrieNode
    { nodeDictId :: Word16
    , nodeChildren :: Map Word8 Int
    }
    deriving (Eq, Show)

-- | Immutable cursor over an arena-backed byte trie.
data TrieCursor = TrieCursor
    { cursorNodes :: IntMap TrieNode
    , cursorCurrent :: Int
    , cursorNextNode :: Int
    }
    deriving (Eq, Show)

-- | Maximum dictionary size supported by the 16-bit wire index.
defaultMaxDictionarySize :: Int
defaultMaxDictionarySize = 65536

-- | An empty trie whose root represents dictionary index zero.
emptyCursor :: TrieCursor
emptyCursor =
    TrieCursor
        { cursorNodes = IntMap.singleton 0 (TrieNode 0 Map.empty)
        , cursorCurrent = 0
        , cursorNextNode = 1
        }

-- | Follow one byte edge, returning 'Nothing' when the edge is absent.
stepCursor :: Word8 -> TrieCursor -> Maybe TrieCursor
stepCursor byte cursor = do
    node <- IntMap.lookup (cursorCurrent cursor) (cursorNodes cursor)
    child <- Map.lookup byte (nodeChildren node)
    pure cursor{cursorCurrent = child}

-- | Add a byte edge at the current node. The cursor does not advance.
insertCursor :: Word8 -> Word16 -> TrieCursor -> TrieCursor
insertCursor byte newDictId cursor =
    cursor
        { cursorNodes =
            IntMap.insert newNodeIndex newNode
                (IntMap.adjust addChild (cursorCurrent cursor) (cursorNodes cursor))
        , cursorNextNode = newNodeIndex + 1
        }
  where
    newNodeIndex = cursorNextNode cursor
    newNode = TrieNode newDictId Map.empty
    addChild node = node{nodeChildren = Map.insert byte newNodeIndex (nodeChildren node)}

-- | Return a cursor to the trie root without changing its dictionary.
resetCursor :: TrieCursor -> TrieCursor
resetCursor cursor = cursor{cursorCurrent = 0}

-- | Dictionary ID at the current cursor position.
cursorDictId :: TrieCursor -> Word16
cursorDictId cursor = maybe 0 nodeDictId (IntMap.lookup (cursorCurrent cursor) (cursorNodes cursor))

-- | Whether the cursor currently points at the trie root.
cursorAtRoot :: TrieCursor -> Bool
cursorAtRoot cursor = cursorCurrent cursor == 0

-- | Encode bytes as LZ78 tokens, capping the dictionary at the requested size.
encode :: ByteString -> Int -> [Token]
encode input maxDictionarySize = reverse (finish finalCursor reversedTokens)
  where
    cappedSize = max 1 (min defaultMaxDictionarySize maxDictionarySize)
    (finalCursor, _, reversedTokens) = BS.foldl' encodeByte (emptyCursor, 1, []) input

    encodeByte (cursor, nextId, tokens) byte =
        case stepCursor byte cursor of
            Just advanced -> (advanced, nextId, tokens)
            Nothing ->
                let token = Token (cursorDictId cursor) byte
                    canInsert = nextId < cappedSize
                    inserted =
                        if canInsert
                            then insertCursor byte (fromIntegral nextId) cursor
                            else cursor
                    followingId = if canInsert then nextId + 1 else nextId
                 in (resetCursor inserted, followingId, token : tokens)

    finish cursor tokens
        | cursorAtRoot cursor = tokens
        | otherwise = Token (cursorDictId cursor) 0 : tokens

-- | Decode tokens, optionally enforcing and trimming to an original length.
decode :: [Token] -> Maybe Int -> Either Lz78Error ByteString
decode tokens originalLength = do
    validateLength originalLength
    (outputReversed, _, _) <- foldTokens tokens
    let output = BS.pack (reverse outputReversed)
        trimmed = maybe output (`BS.take` output) originalLength
        actualLength = BS.length trimmed
    case originalLength of
        Just expected
            | actualLength /= expected -> Left (DecodedLengthMismatch expected actualLength)
        _ -> Right trimmed
  where
    validateLength (Just value)
        | value < 0 = Left (InvalidOriginalLength value)
    validateLength _ = Right ()

    foldTokens = go (IntMap.singleton 0 (0, 0)) 1 [] 0

    go table _ output outputLength [] = Right (output, outputLength, table)
    go table nextId output outputLength (token : rest) = do
        prefix <- reconstruct table nextId (dictIndex token)
        let withPrefix = reverse prefix ++ output
            prefixLength = outputLength + length prefix
            includeNext = maybe True (prefixLength <) originalLength
            withNext = if includeNext then nextChar token : withPrefix else withPrefix
            nextLength = if includeNext then prefixLength + 1 else prefixLength
            nextTable = IntMap.insert nextId (dictIndex token, nextChar token) table
        go nextTable (nextId + 1) withNext nextLength rest

-- | Rebuild one dictionary entry by following its parent chain.
reconstruct :: IntMap (Word16, Word8) -> Int -> Word16 -> Either Lz78Error [Word8]
reconstruct table nextId index = walk index []
  where
    walk 0 bytes = Right bytes
    walk current bytes =
        case IntMap.lookup (fromIntegral current) table of
            Nothing -> Left (InvalidDictionaryIndex current nextId)
            Just (parent, byte) -> walk parent (byte : bytes)

-- | Serialize tokens using the CMP01 big-endian fixed-width wire format.
serialiseTokens :: [Token] -> Int -> ByteString
serialiseTokens tokens originalLength =
    BS.pack
        ( word32Bytes (fromIntegral originalLength)
            ++ word32Bytes (fromIntegral (length tokens))
            ++ concatMap tokenBytes tokens
        )
  where
    tokenBytes token = word16Bytes (dictIndex token) ++ [nextChar token, 0]

-- | Parse and validate a complete CMP01 wire stream.
deserialiseTokens :: ByteString -> Either Lz78Error ([Token], Int)
deserialiseTokens input
    | actualLength < 8 = Left (HeaderTooShort actualLength)
    | actualLength /= expectedLength = Left (WireLengthMismatch expectedLength actualLength)
    | otherwise = do
        tokens <- traverse readToken [0 .. tokenCount - 1]
        Right (tokens, originalLength)
  where
    actualLength = BS.length input
    originalLength = fromIntegral (readWord32 input 0)
    tokenCount = fromIntegral (readWord32 input 4)
    expectedLength = 8 + tokenCount * 4

    readToken tokenNumber =
        let offset = 8 + tokenNumber * 4
            reserved = BS.index input (offset + 3)
         in if reserved /= 0
                then Left (NonZeroReservedByte tokenNumber reserved)
                else
                    Right
                        Token
                            { dictIndex = readWord16 input offset
                            , nextChar = BS.index input (offset + 2)
                            }

-- | Encode and serialize bytes with an explicit dictionary cap.
compress :: ByteString -> Int -> ByteString
compress input maxDictionarySize =
    serialiseTokens (encode input maxDictionarySize) (BS.length input)

-- | Encode and serialize bytes with the 65,536-entry default dictionary.
compressDefault :: ByteString -> ByteString
compressDefault input = compress input defaultMaxDictionarySize

-- | Parse and decode a complete CMP01 wire stream.
decompress :: ByteString -> Either Lz78Error ByteString
decompress input = do
    (tokens, originalLength) <- deserialiseTokens input
    decode tokens (Just originalLength)

word16Bytes :: Word16 -> [Word8]
word16Bytes value =
    [ fromIntegral (value `shiftR` 8)
    , fromIntegral value
    ]

word32Bytes :: Word32 -> [Word8]
word32Bytes value =
    [ fromIntegral (value `shiftR` 24)
    , fromIntegral (value `shiftR` 16)
    , fromIntegral (value `shiftR` 8)
    , fromIntegral value
    ]

readWord16 :: ByteString -> Int -> Word16
readWord16 input offset =
    (fromIntegral (BS.index input offset) `shiftL` 8)
        .|. fromIntegral (BS.index input (offset + 1))

readWord32 :: ByteString -> Int -> Word32
readWord32 input offset =
    (fromIntegral (BS.index input offset) `shiftL` 24)
        .|. (fromIntegral (BS.index input (offset + 1)) `shiftL` 16)
        .|. (fromIntegral (BS.index input (offset + 2)) `shiftL` 8)
        .|. fromIntegral (BS.index input (offset + 3))
